keps/sig-auth/2799-token-controller-deprecation/README.md

```diff
@@ -0,0 +1,395 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a namespace-scoped and service-account-scoped binary label to
+allow users to opt in/out the provision of secret-based service account tokens
+in Token Controller. In addition, it sketches out the phases to deprecate Token
+Controller.
+
+## Motivation
+
+As BoundServiceAccountTokenVolume is GA in 1.22, pods’ service account tokens
+would be obtained via TokenRequest API and stored as projected volume. This
+change obviates the need for auto-generation of secret-based service account
+tokens in Token Controller. The secret-based tokens are [not secure by design](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#background)
+and the token controller is [fragile by design in some cases](https://github.com/kubernetes/kubernetes/issues/98474)
+where it is unable to handle the churns between secrets and service account
+controller loops.
+
+### Goals
+
+- No auto-generation of secret-based service account token.
+- Removal of secret-based service account token that are referenced in service
+  accounts.
+
+### Non-Goals
+
+## Proposal
+
+- In release 1.23, change the service account control loop in Token Controller
+  to not create secret for the service account. At the same time, warn usage of
+  auto-created secret-based service account tokens and enrouage users to use
+  TokenRequest API or manually-created secret-based service account tokens.
+- In release 1.25, purge auto-generated secret-based service account tokens.
+
+### Notes/Constraints/Caveats
+
+- For clusters in upgrade path, users should not upgrade to release 1.25+ unless
+  they are certain of no active usage of secret-based tokens. To consult that
+  information, metric `serviceaccount_stale_tokens_total` or audit annotation
+  `authentication.k8s.io/stale-token` could be used.
+- A warning mechanism should be implemented to push users to migrate and it
+  will exist for at least one year before release N+Y.
+
+### Risks and Mitigations
+
+- In release 1.23, there is only risk in implementation which would be mitigated by
+  tests.
+- In release 1.25, usage of secret-based token that are referenced in service
+  accounts might stop working. To mitigate, users can use TokenRequest API or
+  manually created tokens.
+
+## Design Details
+
+### Token Controller
+
+1.  Token Controller stops to create secret for service accounts.
+2.  Token Controller starts to remove secrets that are not referenced in service
```

Comment:
This needs significantly more detail around how we will make it safe, especially in detecting use of auto-generated secret tokens outside of pods, marking of those secrets to make that use visible, and avoiding breaking those.

My initial recommendation for this would be:
1. in the 1.24 timeframe (at which point all pod-based usage of autogenerated tokens should have dried up)
    1. annotate all autogenerated secrets (token secrets referenced from service account's `secrets` field) with a date indicating when we started tracking use (e.g. "kubernetes.io/tracked-since: $date")
    2. when a legacy token is used, if it is from a secret with the tracked-since annotation, issue a warning, bump a metric, mark the secret as being used (probably with a label), and maaaybe set a last-used date on the secret (need to think about the performance impact of this... if we expect low usage, and the write happens at most once per day, this might be ok)
3. in a later release (1.25? 1.26?), start deleting autogenerated tokens IFF they do not have the "being used" label and have been tracked for a sufficient period of time (X months? 1 year? maybe default to 1 year and give the admin control to lengthen or shorten that?)

That does a few things:
* makes usage of the secrets visible (can select them with a label selector to find them, can see when tokens were last used)
* lets us decide how long to monitor for use before removing
* keeps upgrades non-disruptive ("we won't remove existing tokens that are in use")

Reply:
I like this - "info about scope of breakage / ability to warn about possible breakage" before we decide to break anything.

Reply:
in-use label has to be revokable so that we can clean up tokens that haven't been used for a while otherwise we can only cleanup never-used tokens.